lib/Sema/NameBinding.cpp

```diff
@@ -36,60 +37,539 @@
 using namespace swift;
 
 //===----------------------------------------------------------------------===//
-// NameBinder
+// MARK: NameBinder and supporting types
 //===----------------------------------------------------------------------===//
 
 using ImportedModule = ModuleDecl::ImportedModule;
+using ImportedModuleDesc = SourceFile::ImportedModuleDesc;
 using ImportOptions = SourceFile::ImportOptions;
+using ImportFlags = SourceFile::ImportFlags;
+
+namespace {
+  /// Represents an import which the NameBinder knows exists, but which has not
+  /// yet had its options checked, module loaded, or cross-imports found.
+  ///
+  /// An UnboundImport may represent a physical ImportDecl written in the
+  /// source, or it may represent a cross-import overlay that has been found and
+  /// needs to be loaded.
+  struct UnboundImport {
+    /// If this UnboundImport directly represents an ImportDecl, the ImportDecl
+    /// it represents.
+    ///
+    /// This property should only be used to control where information is added
+    /// to the AST. Don't retrieve information directly from the ImportDecl; use
+    /// the other member variables.
+    NullablePtr<ImportDecl> ID;
+
+    SourceLoc importLoc;
+    ImportOptions options;
+    StringRef privateImportFileName;
+    Located<ImportKind> importKind;
+    ModuleDecl::AccessPathTy modulePath;
+    ModuleDecl::AccessPathTy declPath;
+
+    NullablePtr<DeclAttributes> attrs;
```

Comment:
Bit confused here -- why do we need both `DeclAttributes` and `ImportOptions`? Is `ImportOptions` a parsed/cached representation primarily present to avoid reparsing the attributes?

Reply:
`options` is always set; `attrs` is only set on `UnboundImport`s which directly correspond to an `ImportDecl` (not ones for cross-import overlays). When we validate `options`, we will use `attrs` if available to provide fix-its removing the invalid attributes.

In other words, `options` is the source of truth for binding the import; `attrs` is used if available to emit better diagnostics.

I should explain this more explicitly.

Reply:
I reworked this whole thing so that `attrs` is no longer a separate property and `ID` and `underlyingModule` are merged into a single `PointerUnion` (since we will have exactly one of the two at any given time).